src/java/com/zimbra/oauth/handlers/impl/FacebookContactsImport.java

```diff
@@ -0,0 +1,569 @@
+     * @return Json contacts response
+     * @throws ServiceException If there are issues retrieving the data
+     * @throws IOException If there are issues executing the request
+     */
+    protected JsonNode getContactsRequest(String url) throws ServiceException, IOException {
+
+        final GetMethod get = new GetMethod(url);
+        ZimbraLog.extensions.trace("Fetching contacts for import.");
+        return OAuth2Handler.executeRequestForJson(get);
+    }
+
+    @Override
+    public void importData(List<Integer> folderIds, boolean fullSync) throws ServiceException {
+        final String refreshToken = refresh();
+        final Mailbox mailbox = mDataSource.getMailbox();
+        final int folderId = mDataSource.getFolderId();
+        // existing contacts from the datasource folder
+        final Set<String> existingContacts = getExistingContacts(mailbox,
+                folderId, A_imAddress1);
+
+        String respContent = "";
+        String nextPageUrl = null;
+        try {
+            do {
+                // build contacts url, query params with access token or use Facebook nextPageUrl if defined
+                final String url = buildContactsUrl(nextPageUrl, refreshToken);
+                nextPageUrl = null;
+                // log this only at the most verbose level, because this contains
+                // privileged information
+                ZimbraLog.extensions.debug(
+                        "Attempting to sync Facebook contacts. URL: %s", url);
+                final JsonNode jsonResponse = getContactsRequest(url);
+                respContent = jsonResponse.toString();
+                if (jsonResponse.has("error") && jsonResponse.get("error").isContainerNode()) {
+                    ZimbraLog.extensions.debug(
+                            "Error found in JSON response object while fetching contacts. Response body: %s",
+                            respContent);
+                    throw ServiceException.FAILURE("An error was found in the api response. "
+                            + jsonResponse.get("error").asText(""), null);
+                }
+                // log this only at the most verbose level, because this contains
+                // privileged information
+                if (jsonResponse != null && jsonResponse.isContainerNode()) {
+                    if (jsonResponse.has("data")
+                            && jsonResponse.get("data").isContainerNode()) {
+                        final JsonNode contactsObject = jsonResponse.get("data");
+                        handleApiResponse(existingContacts, contactsObject);
+                        // check for next page
+                        if (jsonResponse.has("paging")
+                                && jsonResponse.get("paging").has("next")) {
+                            nextPageUrl = jsonResponse.get("paging").get("next").asText();
+                        }
+                    } else {
+                        ZimbraLog.extensions
```

Comment:
This will be info logged a lot for every user's sync that doesn't have any new contacts.

Reply:
This will happen only if the required facebook 'data' node is missing which is expected on all responses to this api call.